_posts/2017-10-09-functional-calisthenics.md

```diff
@@ -0,0 +1,123 @@
+
+## Name everything
+Most basic one. It will seem harsh for experienced functional programmers. Fisrts one to be broken on production code.
+
+### Description
+All functions need to be named. Which means that you shouldn't use lambdas or anonymous functions. Also, name all your variables as per standard clean code rules.
+
+### Expected Outcome
+**Learn to recognize patterns on your code.**
+It will help recognizing signatures that repeat, and the applying of DRY on your code.
+
+## No mutable state
+This is the basis of FP.
+
+### Description
+You shouldn't using any variable of any type that can mutate.
+
+### Expected Outcome
+**Learn how to create code around immutable variables.**
+Two main benefits: 
+    - FP is about immutability. Most of its benefits comes from the fact that all your functions are (or should be) referencially transparent. 
+    - Use of recursion as your main looping technique. No for or while loops in your code. For comprehensions abide by this rule.
+
+## Exhaustive conditionals
+This rule is mostly preparation work.
+
+### Description
+There can not be an if without an else. Switches and pattern matching should always have all paths considered (either through default paths or because all options have been considered).
+
+### Expected Outcome
+**All possible paths have been taken into account.**
+Is a know cause of issues. And also helps preparing for the 5th rule.
+
+## Do not use intermediate variables
+### Description
+There shouldn't be any variable declared in the body of a function. Only parameters and other functions should appear on the body.
+
+### Expected Outcome
+**Use and understanding of functional pipelines. Starting on the path of functional composition.**
+Functions are the building blocks of functional languages and the combination of functions create the logic of your application. Not having intermediate variables means changing the way that you think about your functions, and how the body of a function (composed of other functions) is, so the code remains legible and easy to follow.
+
+## Expressions not statements.
+A consequence of the two previous rule. Or a generalization.
+
+### Description
+All lines should be expressions. That's it, all lines should return a value.
+
+### Expected Outcome
+**Full purity on the code.**
+Statements that are not expression are there for their side-effects. Nothing like that should be in the code.
+
+## No Explicit recursion
+Not as harsh as it looks.
+
+### Description
+You shouldn't use explicit recursion like Clojure's `loop/recur` forms or F# `let rec` form.
+
+### Expected Outcome
+**Learning the power of map and reduce.**
```

Comment:
I would probably add that when you do recursion you are missing a HOF that will take the low level recursion out of the way.